src/wallet3/db_schema.cpp

```diff
@@ -420,5 +425,41 @@ namespace wallet
   {
     return prepared_get<int64_t>("SELECT output_count FROM metadata WHERE id=0;");
   }
+  void
+  WalletDB::save_keys(const std::string& spend_priv_str, const std::string& spend_pub_str, const std::string& view_priv_str, const std::string& view_pub_str)
+  {
+    std::string query = "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0";
+
+    auto st = prepared_st(query);
+    st->executeStep();
+    auto from_db = db::get<std::string, std::string, std::string, std::string>(st);
+    const auto loaded_spend_priv_str = std::get<0>(from_db);
+    const auto loaded_spend_pub_str = std::get<1>(from_db);
+    const auto loaded_view_priv_str = std::get<2>(from_db);
+    const auto loaded_view_pub_str = std::get<3>(from_db);
+
+    if ((loaded_spend_priv_str != "" && loaded_spend_priv_str != spend_priv_str) ||
+        (loaded_spend_pub_str != "" && loaded_spend_pub_str != spend_pub_str) ||
+        (loaded_view_priv_str != "" && loaded_view_priv_str != view_priv_str) ||
+        (loaded_view_pub_str != "" && loaded_view_pub_str != view_pub_str))
+          throw std::runtime_error("loaded keys do not match database file");
+
+    prepared_exec("UPDATE metadata SET spend_priv = ?, spend_pub = ?, view_priv = ?, view_pub = ? where id = 0;",
+        spend_priv_str,
+        spend_pub_str,
+        view_priv_str,
+        view_pub_str);
```

Comment:
I think these should be BLOBs (bytes) rather than hex strings in the database.

Reply:
Primarily it's a cleaner database design because the usable type here is a byte buffer, not a hex string, so storing it as hex where you always have to do an encode/decode step when going to/from the database means the database isn't storing "the data", it's storing a serialized form of the data.

As for human readability, the sqlite hex() function solves this:
`SELECT hex(somecol) FROM sometable`